packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -0,0 +1,129 @@
+ * Licensed under the MIT License.
+ */
+
+import { EventEmitter } from "events";
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import { IConnectionDetails } from "@fluidframework/container-definitions";
+import { ProtocolOpHandler, Quorum } from "@fluidframework/protocol-base";
+import { ConnectionMode } from "@fluidframework/protocol-definitions";
+import { connectEventName, ConnectionState, ILocalSequencedClient } from "./container";
+
+export class ConnectionStateHandler {
+    private _connectionState = ConnectionState.Disconnected;
+    private _pendingClientId: string | undefined;
+    private _clientId: string | undefined;
+
+    public get connectionState(): ConnectionState {
+        return this._connectionState;
+    }
+
+    public get connected(): boolean {
+        return this.connectionState === ConnectionState.Connected;
+    }
+
+    public get clientId(): string | undefined {
+        return this._clientId;
+    }
+
+    public get pendingClientId(): string | undefined {
+        return this._pendingClientId;
+    }
+
+    constructor(
+        private readonly protocolHandler: () => ProtocolOpHandler | undefined,
+        private readonly logConnectionStateChangeTelemetry:
+            (value: ConnectionState, oldState: ConnectionState, reason?: string | undefined) => void,
+        private readonly propagateConnectionState: () => void,
+        private readonly isContainerLoaded: () => boolean,
+        private readonly logger: ITelemetryLogger,
+    ) {
+    }
+
+    public receivedAddMemberEvent(clientId: string, quorum: Quorum) {
+        // This is the only one that requires the pending client ID
+        if (clientId === this.pendingClientId) {
+            this.setConnectionState(ConnectionState.Connected);
+        }
+    }
+
+    public receivedDisconnectEvent(reason: string) {
+        this.setConnectionState(ConnectionState.Disconnected, reason);
+    }
+
+    public receivedConnectEvent(
+        emitter: EventEmitter,
+        connectionMode: ConnectionMode,
+        details: IConnectionDetails,
+        opsBehind?: number,
+    ) {
+        const oldState = this._connectionState;
+        this._connectionState = ConnectionState.Connecting;
+
+        // Stash the clientID to detect when transitioning from connecting (socket.io channel open) to connected
+        // (have received the join message for the client ID)
+        // This is especially important in the reconnect case. It's possible there could be outstanding
+        // ops sent by this client, so we should keep the old client id until we see our own client's
+        // join message. after we see the join message for out new connection with our new client id,
+        // we know there can no longer be outstanding ops that we sent with the previous client id.
+        this._pendingClientId = details.clientId;
+
+        emitter.emit(connectEventName, opsBehind);
```

Comment:
I'd remove it from here and leave it on container - less arguments to pass in, and not raising events on some object.
Is the concern that we did not change state before raising an event?
Maybe we can do it after logic in here run through?

Reply:
We could not emit after the logic because we could move to Connected state if read mode etc. Emitting before, I was concerned about not setting the state before emitting. Although it is synchronous but do not if we would have any implecations.

Reply:
I raised an event on handler class which is then caught by container and then emitted by container itself.